op-batcher/txmgr.go

```diff
@@ -2,121 +2,153 @@ package op_batcher
 
 import (
 	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
 	"time"
 
 	"github.com/ethereum-optimism/optimism/op-proposer/txmgr"
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
 )
 
-func (l *BatchSubmitter) submitTransaction(data []byte) error {
-	// Query for the submitter's current nonce.
-	ctx, cancel := context.WithTimeout(l.ctx, time.Second*10)
-	nonce, err := l.cfg.L1Client.NonceAt(ctx, l.addr, nil)
-	cancel()
-	if err != nil {
-		l.log.Error("unable to get current nonce", "err", err)
-		return err
+const networkTimeout = 2 * time.Second // How long a single network request can take. TODO: put in a config somewhere
+
+// TransactionManager wraps the simple txmgr package to make it easy to send & wait for transactions
+type TransactionManager struct {
+	// Config
+	batchInboxAddress common.Address
+	senderAddress     common.Address
+	chainID           *big.Int
+	// Outside world
+	txMgr    txmgr.TxManager
+	l1Client *ethclient.Client
+	signerFn func(types.TxData) (*types.Transaction, error)
+	log      log.Logger
+}
+
+func NewTransactionManger(log log.Logger, txMgrConfg txmgr.Config, batchInboxAddress common.Address, chainID *big.Int, privKey *ecdsa.PrivateKey, l1Client *ethclient.Client) *TransactionManager {
+	signerFn := func(rawTx types.TxData) (*types.Transaction, error) {
+		return types.SignNewTx(privKey, types.LatestSignerForChainID(chainID), rawTx)
 	}
 
-	// Create the transaction
-	ctx, cancel = context.WithTimeout(l.ctx, time.Second*10)
-	tx, err := l.CraftTx(ctx, data, nonce)
-	cancel()
-	if err != nil {
-		l.log.Error("unable to craft tx", "err", err)
-		return err
+	t := &TransactionManager{
+		batchInboxAddress: batchInboxAddress,
+		senderAddress:     crypto.PubkeyToAddress(privKey.PublicKey),
+		chainID:           chainID,
+		txMgr:             txmgr.NewSimpleTxManager("batcher", txMgrConfg, l1Client),
+		l1Client:          l1Client,
+		signerFn:          signerFn,
+		log:               log,
 	}
+	return t
+}
 
-	// Construct the a closure that will update the txn with the current gas prices.
+// SendTransaction creates & submits a transaction to the batch inbox address with the given `data`.
+// It currently uses the underlying `txmgr` to handle transaction sending & price management.
+// This is a blocking method. It should not be called concurrently.
+// TODO: where to put concurrent transaction handling logic.
```

Comment:
I feel like the logic should live in an implementation of the `TxManager` so that it can be dropped into any service, like instead of `txmgr.NewSimpleTxManager`